libcasr/src/java.rs

```diff
@@ -0,0 +1,272 @@
+            .extend_from_slice(&last_block.body.iter().rev().copied().collect::<Vec<&str>>());
+
+        for block in blocks.iter().rev().skip(1) {
+            if let Some(cur_num) = block.footer {
+                let diff = prev_num - cur_num;
+                forward_stacktrace.extend_from_slice(
+                    &block.body[..diff]
+                        .iter()
+                        .rev()
+                        .copied()
+                        .collect::<Vec<&str>>(),
+                );
+                backward_stacktrace.extend_from_slice(&block.body[diff..]);
+                prev_num = cur_num;
+            } else {
+                forward_stacktrace.extend_from_slice(
+                    &block.body[..prev_num]
+                        .iter()
+                        .rev()
+                        .copied()
+                        .collect::<Vec<&str>>(),
+                );
+                backward_stacktrace.extend_from_slice(&block.body[prev_num..]);
+                break;
+            }
+        }
+
+        backward_stacktrace.reverse();
+        backward_stacktrace.extend_from_slice(forward_stacktrace.as_slice());
+        if forward_stacktrace.is_empty() {
+            return Err(Error::Casr("Empty stacktrace.".to_string()));
+        }
+        *JAVA_CRASH_LINE.lock().unwrap() = forward_stacktrace[0].to_string();
+
+        Ok(backward_stacktrace.iter().map(|x| x.to_string()).collect())
+    }
+
+    fn parse_stacktrace(entries: &[String]) -> Result<Stacktrace> {
+        let mut stacktrace = Stacktrace::new();
+        let re = Regex::new(r"(?:\s|\t)*at (.*)\((.*)\)").unwrap();
+        let rd = Regex::new(r"(.*):(\d+)").unwrap();
+
+        for entry in entries.iter() {
+            let Some(cap) = re.captures(entry) else {
+                return Err(Error::Casr(format!(
+                    "Couldn't parse stacktrace line: {entry}")
+                ));
+            };
+            let debug = cap.get(2).unwrap().as_str().to_string();
+            if debug.contains("Unknown Source") || debug.contains("Native Method") {
+                continue;
+            }
+            let mut stentry = StacktraceEntry::default();
+            stentry.debug.file = if let Some(decap) = rd.captures(&debug) {
+                stentry.debug.line = decap.get(2).unwrap().as_str().parse::<u64>().unwrap();
+                decap.get(1).unwrap().as_str().to_string()
+            } else {
+                debug
+            };
+            stentry.function = cap.get(1).unwrap().as_str().to_string();
+
+            stacktrace.push(stentry);
+        }
+        Ok(stacktrace)
+    }
+}
+
+/// Structure provides an interface for parsing java exception message.
+pub struct JavaException;
+
+impl Exception for JavaException {
+    fn parse_exception(description: &str) -> Option<ExecutionClass> {
+        let re = Regex::new(
+            r"(?:Caused by: |Exception in thread .*? |== Java Exception: )(?:(\S+?): |)(\S+)?",
```

Comment:
why do we need `?` after `.*`?

Reply:
And after `\S+`?

Reply:
`.*?` will capture all before first space or `:`